toxcore/nat_traversal.c

```diff
@@ -0,0 +1,198 @@
+/* nat_traversal.c -- Functions to traverse a NAT (UPnP, NAT-PMP).
+ *
+ *  Copyright (C) 2016 Tox project All Rights Reserved.
+ *
+ *  This file is part of Tox.
+ *  Tox is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  Tox is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with Tox.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#ifdef HAVE_LIBMINIUPNPC
+#include <stdio.h>
+#include <miniupnpc/miniupnpc.h>
+#include <miniupnpc/miniwget.h>
+#include <miniupnpc/upnpcommands.h>
+#include <miniupnpc/upnperrors.h>
+#endif
+
+#ifdef HAVE_LIBNATPMP
+#include <unistd.h>
+#include <natpmp.h>
+#endif
+
+#include "nat_traversal.h"
+
+#define UNUSED(x) (void)(x)
+
+#ifdef HAVE_LIBMINIUPNPC
+/* Setup port forwarding using UPnP */
+void upnp_map_port(Logger *log, NAT_TRAVERSAL_PROTO proto, uint16_t port)
+{
+    int error;
+    struct UPNPDev *devlist = NULL;
+
+    LOGGER_DEBUG(log, "Attempting to set up UPnP port forwarding");
+
+#if MINIUPNPC_API_VERSION < 14
+    devlist = upnpDiscover(1000, NULL, NULL, 0, 0, &error);
+#else
+    devlist = upnpDiscover(1000, NULL, NULL, 0, 0, 2, &error);
+#endif
+
+    if (error) {
+        LOGGER_WARNING(log, "UPnP discovery failed (%s)", strupnperror(error));
+        return;
+    }
+
+    struct UPNPUrls urls;
+
+    struct IGDdatas data;
+
+    char lanaddr[64];
+
+    error = UPNP_GetValidIGD(devlist, &urls, &data, lanaddr, sizeof(lanaddr));
+
+    freeUPNPDevlist(devlist);
+
+    switch (error) {
+        case 0:
+            LOGGER_WARNING(log, "No IGD was found.");
+            break;
+
+        case 1:
+            LOGGER_INFO(log, "A valid IGD has been found.");
+
+            char portstr[10];
+            snprintf(portstr, sizeof(portstr), "%d", port);
+
+            switch (proto) {
+                case NAT_TRAVERSAL_UDP:
+                    error = UPNP_AddPortMapping(urls.controlURL, data.first.servicetype, portstr, portstr, lanaddr, "Tox", "UDP", 0, "0");
+                    break;
+
+                case NAT_TRAVERSAL_TCP:
+                    error = UPNP_AddPortMapping(urls.controlURL, data.first.servicetype, portstr, portstr, lanaddr, "Tox", "TCP", 0, "0");
```

Comment:
This line is exactly equal to the one in the UDP branch, apart from the "TCP" string. You can factor out the common code and use some "nat_traversal_name" function or just assign some variable. Perhaps use NULL in the unknown proto case.